python/ray/data/_internal/output_buffer.py

```diff
@@ -73,16 +97,14 @@ def next(self) -> Block:
         block_to_yield = self._buffer.build()
         block_remainder = None
         block = BlockAccessor.for_block(block_to_yield)
-        if (
-            block.size_bytes()
-            >= MAX_SAFE_BLOCK_SIZE_FACTOR * self._target_max_block_size
-        ):
-            # Slice a block to respect the target max block size.  We only do
-            # this if we are more than 50% above the target block size, because
-            # this ensures that the last block produced will be at least half
-            # the block size.
-            num_bytes_per_row = block.size_bytes() // block.num_rows()
-            target_num_rows = max(1, self._target_max_block_size // num_bytes_per_row)
+
+        if self._block_row_limit(block) or self._block_size_limit(block):
+            num_bytes_per_row = max(1, block.size_bytes() // block.num_rows())
+            target_num_rows_by_size = self._target_max_block_size // num_bytes_per_row
+            target_num_rows_by_rows = (
+                self._target_max_rows_per_block or block.num_rows()
+            )
+            target_num_rows = min(target_num_rows_by_size, target_num_rows_by_rows)
```

Comment:
To simplify the logic, I think we can just consider one factor a time. I.E., only one factor can be non-None.

Reply:
+1